src/storage.cpp

```diff
@@ -1061,32 +1068,34 @@ namespace libtorrent
 			return piece_manager::fatal_disk_error;
 		}
 
-		m_save_path.swap(save_path);
-		// now we have old save path in save_path
+		std::string old_save_path;
+		m_save_path.swap(old_save_path);
 
 		std::set<std::string> subdirs;
 		for (i = 0; i < f.num_files(); ++i)
 		{
 			// files moved out to absolute paths are not moved
-			if (is_complete(f.file_path(i))) continue;
+			if (f.file_absolute_path(i)) continue;
 
 			if (has_parent_path(f.file_path(i)))
 				subdirs.insert(parent_path(f.file_path(i)));
-			std::string old_path = combine_path(save_path, f.file_path(i));
 
-			// we may still have some files in old save_path
+			std::string const old_path = combine_path(old_save_path, f.file_path(i));
+
+			// we may still have some files in old old_save_path
 			// eg. if (flags == dont_replace && exists(new_path))
 			// ignore errors when removing
 			error_code ignore;
 			remove(old_path, ignore);
 		}
 
-		for (std::string subdir: subdirs)
+		for (std::string const& s : subdirs)
 		{
-			e.clear();
-			while (!subdir.empty() && !e)
+			error_code err;
+			std::string subdir = combine_path(old_save_path, s);
+			while (subdir != old_save_path && !err)
```

Comment:
combine_path isn't though. this saves a call to that inside the loop. the comparison here will most likely be quite fast too, since the strings have different length (that should be an early exit)